### ISSUE TITLE
install-apks not completely compatible with MissingSplitsManagerFactory

### ISSUE THREAD

**Issue by malachid** [google/bundletool#109]
**Describe the bug**
In an attempt to prevent crashes when a user sideloads an incorrect version of the application, we are attempting to use [the instructions](https://developer.android.com/guide/app-bundle/sideload-check?hl=en-419) to call `MissingSplitsManagerFactory.create(this).disableAppIfMissingRequiredSplits()`.

Unfortunately, this call only works if _all_ splits are missing; not if some splits are missing.

Alternatively, the tooling could not allow an invalid version to be sideloaded.

**Bundletool version(s) affected**
Version: 0.10.2

**Stacktrace**
The tooling does not crash. Instead, later native or resource usage crashes due to being missing.

**To Reproduce**
1. Include a bundle configuration in your app including two splits
```
    bundle {
        language {
            enableSplit = false
        }
        density {
            enableSplit = true
        }
        abi {
            enableSplit = true
        }
    }
```
2. Follow [these instructions](https://developer.android.com/guide/app-bundle/sideload-check?hl=en-419) to call `MissingSplitsManagerFactory.create(this).disableAppIfMissingRequiredSplits()`
3. Use `bundletool get-device-spec` to get the specification to a high-res arm64 device.
4. Use `bundletool build-apks` to generate a device-specific apks
5. Use `bundletool install-apks` to install the generated apks on different devices (32-bit high-res, 32-bit low-res, 64-bit low-res, etc)
6. Launch the app

**Expected behavior**
I expected install-apks to notify the user that the install failed.  Short of that, I expected disableAppIfMissingRequiredSplits to return true if any of the splits were missing.

**Actual behavior**
Running `adb shell dumpsys package YOUR_APP | grep split` will tell you if disableAppIfMissingRequiredSplits will fail.
1. If dumpsys returns just `[base]`, the reinstall dialog is launched
2. If dumpsys shows both splits `[base, config.x86, config.xxhdpi]`(etc), the splits are not missing and it will work correctly
3. If dumpsys shows EITHER `[base, config.x86]` or `[base, config.xxhpi]`, but not the other split -- disableAppIfMissingRequiredSplits will incorrectly return false and let the app continue with missing native/resources.

**Known workaround**
Instead of doing:
```
if (MissingSplitsManagerFactory.create(this).disableAppIfMissingRequiredSplits()) {
    return
}
```
you have to do:
```
if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
    try {
        val splits: Array<String> = packageManager.getPackageInfo(packageName, 0).splitNames ?: arrayOf()
        android.util.Log.d(TAG, "Splits found: ${splits.joinToString()}")
        if (MissingSplitsManagerFactory.create(this).disableAppIfMissingRequiredSplits()) {
            android.util.Log.d(TAG, "Missing splits. aborting.")
            return
        }
        if (splits.size != 2) {
            android.util.Log.d(TAG, "MissingSplitsManagerFactory failed to trigger. Aborting.")
            /**
             * FirebaseRemoteParameters will sometimes crash going this route, causing
             * the ANR-like UI; however it then shows the correct Reinstall link.  Hopefully,
             * all of this can be removed once disableAppIfMissingRequiredSplits works properly.
             */
            packageManager.setComponentEnabledSetting(
                    ComponentName(this, PlayCoreMissingSplitsActivity::class.java),
                    COMPONENT_ENABLED_STATE_ENABLED,
                    0)
            
            (getSystemService(ACTIVITY_SERVICE) as ActivityManager)
                    .appTasks
                    ?.asSequence()
                    ?.forEach { it.finishAndRemoveTask() }

            startActivity(Intent(this, PlayCoreMissingSplitsActivity::class.java).addFlags(
                    FLAG_ACTIVITY_SINGLE_TOP +
                            FLAG_ACTIVITY_NEW_TASK +
                            FLAG_ACTIVITY_CLEAR_TOP +
                            FLAG_ACTIVITY_EXCLUDE_FROM_RECENTS +
                            FLAG_ACTIVITY_BROUGHT_TO_FRONT))
            return
        }
        android.util.Log.d(TAG, "No missing splits detected.")
    } catch (ex: PackageManager.NameNotFoundException) {
        android.util.Log.d(TAG, ex.message)
    }
} else {
    android.util.Log.d(TAG, "SplitsManager only works on API21+")
}
```

**Environment:**
OS: Android API21+; tried with multiple OEM and emulator. Consistent results.

**Additional context**
We have had to resort to publishing Fat APKs instead of AABs while attempting to resolve this issue.

**Comment by malachid** [comment]
I forgot to mention, possibly related to:

- https://github.com/google/bundletool/issues/98
- https://github.com/google/bundletool/issues/57

**Comment by plecesne** [comment]
That would technically be a bug for the PlayCore API, but I'll respond here anyway.

The main problem of sideloading is not users sideloading the wrong set of APKs but users sideloading only the base APK, which is why we have focused the API on the latter case which is much simpler to resolve. The former case could also potentially be resolved with the PlayCore API but would require a much bigger investment, and we haven't heard so far from developers that this was a real concern. Have you noticed this happening with real users? Or were you just testing locally the limits of the API?

**Comment by malachid** [comment]
We had enabled AABs in prod, then had a surge of crashes in Fabric due to missing native libraries.  Our assumption was that they were sideloading the 64-bit builds onto 32-bit devices; but we had no way to validate that. We were able to mimic the behavior by doing so.  We are looking to prevent the crash prior to re-enabling AAB in production.

**Comment by plecesne** [comment]
I thought the Android platform did not allow the installation of a 64-bit native library on a 32-bit devices. How did you reproduce this?

**Comment by malachid** [comment]
When running `bundletool install-apks` using an apks generated for a 64-bit device; the tool installs the app, but with the missing native piece.

**Comment by plecesne** [comment]
I suspect bundletool simply did not install the config split with the native library, which may be a bug in bundletool, so the crash would only happen because the native library is not present at all.

**Comment by malachid** [comment]
I agree. That seems to match both dumpsys output and the Device Explorer in AS.

**Comment by plecesne** [comment]
If you need help understanding where this crash comes from, you can reach out to the [Play Developer support](https://support.google.com/googleplay/android-developer/answer/7218994?hl=en), providing as information the package name, versionCode, and the stacktrace you have noticed, and we can help you try and understand what happened.

**Comment by wbervoets** [comment]
I seem to have a few ContentProviders (eg. FirebaseInitProvider, CrashlyticsInitProvider, ...) which still crash the app because they are loaded before my custom app is created (where I do disableAppIfMissingRequiredSplits).

Is there an easy way to make this work?

**Comment by ymakhno** [comment]
> I expected install-apks to notify the user that the install failed.

Bundletool part of the issue is fixed in bundletool 1.4.0.
